CIPs/0017.md

```diff
@@ -0,0 +1,181 @@
+  MaxNonSigners uint // MinQuorumSize()
+}
+
+// All of the data needed by the light client to verify and utilize a proof.
+// This is also crafted to allow for batch verifications using `FirstHashToField`
+// My rough calculations show this struct is ~439 bytes for a non-batched proof with no val set differences.
+LightPlumoProof struct {
+  Proof             []byte // Serialized proof. Latest estimate from @psi = 383 bytes
+  FirstEpoch        uint // First epoch this proof is over - maxNonSigners should be available to client
+  LastEpoch         LightEpochBlock // Last epoch this proof is over
+  VersionNumber     uint // Proof version number, indicates the corresponding verification key to use
+  LastHashToField  []byte  // Hash to field of Last epoch's `snark.EpochBlock` struct. To be used for batch verification.
+                           // If there is no batch verification, this will be empty. May need an additional boolean to signify this.
+                           // If batch verification should be used, newValidators and deletedValidators will be empty.
+  NewValidators     []blsCrypto.SerializedPublicKey // ValSetDiff(FirstEpoch, LastEpoch), added validators
+  DeletedValidators *big.Int // ValSetDiff(FirstEpoch, LastEpoch): bitmap of deleted validators from the `firstEpoch` valset.
+}
+```
+
+The light clients will need to store verification keys corresponding to the existing versions of the snark circuit, as elaborated in #upgradeability. 
+
+There are two proposed extensions for the `les` protocol:
+
+Option 1:
+
+This option involves one extra round trip in order to get peers' available proofs to choose the "best peer". It's seemingly possible to include this extra round trip in the handshake between peers to optimize performance. If the client wants to minimize network traffic and assume proofs are well distributed, they can also bypass the `GetProofInventory` step and just request that the server sends the best proofs it has for the client's current epoch (using the `handleLogic` flag on `GetEpochHeadersOrProofs`).
+```
+GetProofInventory - Asks the server what proofs it has. Server responds with `[]proofMetadata` of all available proofs.
+GetEpochHeadersOrProofs:
+  Client sends (highestEpoch uint, handleLogic boolean, requestedProofs []proofMetadata)
+    `highestEpoch` is the current highestEpoch the client knows of.
+    `handleLogic`: If true, indicates the server should send back the best set of proofs/headers it has
+                   If false, the server should look at `requestedProofs` and send back proofs requested, with headers to fill gaps as appropriate.
+    `requestedProofs`: Set of proofs' metadata requested by the client
+  Server responds with `EpocHeadersOrProofs` dependent on `handleLogic`.
+  Both responses are subject to message size limits limiting the number of proofs/headers.
+EpochHeadersOrProofs: Contains the set of proofs & headers to help the client, as requested in `GetEpochHeadersOrProofs`.
```

Comment:
Based on our discussion yesterday, I would suggest breaking this message into `GetHeaders/Headers` (which exists) and `GetProofs/Proofs` and only using this message set when the client wants explicit control over what is returned. Option two would use a separate message to share it's sync status and respond with a mixture of `Headers` and `Proofs` messages controlled by the server.